app/single_resource/views.py

```diff
@@ -45,24 +57,102 @@ def create():
+        else:  # fields for text descriptors
+            default = None
+            text_association = TextAssociation.query.filter_by(
+                resource_id=resource_id,
+                descriptor_id=descriptor.id
+            ).first()
+            if (text_association is not None):
+                default = text_association.text
+            setattr(SingleResourceForm,
+                    descriptor.name,
+                    StringField(default=default))
     form = SingleResourceForm()
     if form.validate_on_submit():
-        resource.name = form.name.data
-        resource.address = form.address.data
-        resource.latitude = form.latitude.data
-        resource.longitude = form.longitude.data
+        for field_name in resource_field_names[1:]:
+            setattr(resource, field_name, form[field_name].data)
+        add_associations(resource=resource,
+                         form=form,
+                         descriptors=descriptors,
+                         resource_existed=True)
         try:
             db.session.commit()
             flash('Resource updated', 'form-success')
-            return redirect(url_for('single_resource.edit',
-                                    resource_id=resource_id))
+            return redirect(url_for('single_resource.index'))
         except IntegrityError:
             db.session.rollback()
             flash('Error: failed to save resource. Please try again.',
                   'form-error')
-    for field_name in field_names[1:]:
+    for field_name in resource_field_names[1:]:
         form[field_name].data = resource.__dict__[field_name]
     return render_template('single_resource/edit.html',
                            form=form,
                            resource_id=resource_id)
+
+
+def add_associations(resource, form, descriptors, resource_existed=True):
+    for descriptor in descriptors:
+        if (descriptor.values):
+            AssociationClass = OptionAssociation
+            value = int(form[descriptor.name].data)
+            keyword = 'option'
+        else:
+            AssociationClass = TextAssociation
+            value = form[descriptor.name].data
+            keyword = 'text'
+        association = None
+        if (resource_existed):
+            association = AssociationClass.query.filter_by(
+                resource_id=resource.id,
+                descriptor_id=descriptor.id
+            ).first()
+        if (association is not None):
+            setattr(association, keyword, value)
+        else:
+            arguments = {'resource_id': resource.id,
+                         'descriptor_id': descriptor.id,
+                         keyword: value,
+                         'resource': resource,
+                         'descriptor': descriptor}
+            new_association = AssociationClass(**arguments)
+            db.session.add(new_association)
+
+
+@single_resource.route('/<int:resource_id>/delete')
```

Comment:
What do you mean by "it's not public-facing"? Aren't all routes with @login_required not public-facing then?